pkg/sysregistriesv2/system_registries_v2.go

```diff
@@ -519,6 +519,75 @@ func GetRegistries(ctx *types.SystemContext) ([]Registry, error) {
 	return config.Registries, nil
 }
 
+// LocationsForReferences returns a sorted list of locations (i.e., registries)
+// that may be contacted when pulling the specified references.  If the
+// references are empty, all unqualified-search registries and their mirrors
+// will be returned.
+func LocationsForReferences(ctx *types.SystemContext, references []string) ([]string, error) {
+	locations := make(map[string]bool)
+	needUnqualified := len(references) == 0
+
+	addLocation := func(loc string) {
+		locations[loc] = true
+	}
+
+	addRegistry := func(reg *Registry) {
+		if reg == nil {
+			return
+		}
+		addLocation(reg.Location)
+		for _, mir := range reg.Mirrors {
+			addLocation(mir.Location)
+		}
+	}
+
+	for _, ref := range references {
+		// Parse each reference.
+		namedRef, err := reference.ParseNamed(ref)
+		if err != nil {
+			// If it's not a fully-qualified reference, we need to
+			// query unqualified registries and their mirrors.
+			needUnqualified = true
+			continue
+		}
+		// Add the domain of the "raw" reference.
+		addLocation(reference.Domain(namedRef))
+		// Add the registry and all mirrors for the reference.
+		reg, err := FindRegistry(ctx, ref)
+		if err != nil {
+			return nil, err
+		}
+		addRegistry(reg)
+	}
+
+	// If needed, add all unqualified-search registries and their mirrors.
+	if needUnqualified {
+		config, err := getConfig(ctx)
+		if err != nil {
+			return nil, err
+		}
+		for _, searchReg := range config.UnqualifiedSearchRegistries {
+			// Add the search registry.
+			addLocation(searchReg)
+			for _, reg := range config.Registries {
+				if !refMatchesPrefix(searchReg, reg.Prefix) {
```

Comment:
Isn’t this just `FindRegistry(searchReg)`? That doesn’t account for mirrors of namespaces within `searchReg`.

So, this should combine `searchReg` with (each member of) `references` — and c/image does not currently provide any API to qualify unqualified names, I don’t quite want to make it a part of the core API (it should be some opt-in “Podman-like name handling” subpackage), but if we don’t, there’s a risk of the c/image implementation and the Podman implementation getting out of sync.

Reply:
We can't use `FindRegistry(...)` as we need _all_ matching registries (and mirrors) and not only the one with the longest matching prefix. I'll add more verbose comments after changing the API.

Reply:
I thought that we need, for all search registries, the mirrors that apply for qualifying the input name with that registry = the mirrors found by `FindRegistry(combinedName)`.

Is this about the case of `len(references)==0`?